staging/src/k8s.io/component-base/config/serializer/encode.go

```diff
@@ -0,0 +1,103 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serializer
+
+import (
+	"fmt"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+)
+
+// EncodingFormat is a high-level, typed constant configuring how an object should be encoded.
+type EncodingFormat string
+
+const (
+	// ContentTypeYAML encodes a config file in the YAML format.
+	ContentTypeYAML EncodingFormat = "application/yaml"
+	// ContentTypeJSON encodes a config file in compact JSON format.
+	ContentTypeJSON EncodingFormat = "application/json"
+)
+
+// NewStrictYAMLJSONSerializer creates a new implementation of the StrictYAMLJSONSerializer interface.
+func NewStrictYAMLJSONSerializer(scheme *runtime.Scheme) StrictYAMLJSONSerializer {
+	knownEncoders := map[EncodingFormat]runtime.Encoder{}
+	codecs := serializer.NewCodecFactory(scheme)
+	for _, serializerInfo := range codecs.SupportedMediaTypes() {
+		switch serializerInfo.MediaType {
+		case "application/json":
+			knownEncoders[ContentTypeJSON] = serializerInfo.Serializer
+		case "application/yaml":
+			knownEncoders[ContentTypeYAML] = serializerInfo.Serializer
+		}
+	}
+
+	// Create the strict deserializer, and creates a decoder later.
+	strictYAMLSerializer := json.NewStrictYAMLSerializer(json.DefaultMetaFactory, scheme, scheme)
+	universalDecoder := codecs.CodecForVersions(nil, strictYAMLSerializer, nil, runtime.InternalGroupVersioner)
+
+	return &strictYAMLJSONSerializer{scheme, knownEncoders, universalDecoder}
+}
+
+// StrictYAMLJSONSerializer provides encoding/decoding for a config file with YAML or JSON format.
+type StrictYAMLJSONSerializer interface {
+	DecodeInto([]byte, runtime.Object) error
+	Encode(EncodingFormat, schema.GroupVersion, runtime.Object) ([]byte, error)
+}
+
+// strictYAMLJSONSerializer implements the StrictYAMLJSONSerializer interface.
+var _ StrictYAMLJSONSerializer = &strictYAMLJSONSerializer{}
+
+type strictYAMLJSONSerializer struct {
+	scheme        *runtime.Scheme
+	knownEncoders map[EncodingFormat]runtime.Encoder
+	decoder       runtime.Decoder
+}
+
+// DecodeInto decodes bytes into a pointer of the desired type.
+func (cs *strictYAMLJSONSerializer) DecodeInto(data []byte, into runtime.Object) error {
+	out, gvk, err := cs.decoder.Decode(data, nil, into)
+	// return a structured error if the group was registered with the scheme but the version was unrecognized.
+	if gvk != nil && err != nil {
+		if cs.scheme.IsGroupRegistered(gvk.Group) && !cs.scheme.IsVersionRegistered(gvk.GroupVersion()) {
+			return NewUnrecognizedVersionError(*gvk, data)
```

Comment:
this loses error information... will that make it hard to debug the root cause of the failure?

Reply:
is there a reason we're doing fixup of the error here, instead of returning better error information from the decode method?

Reply:
I would have expected to get a notRegisteredErr back from Decode, which would have already indicated the type we asked for didn't exist in the target GVK. If that's not what we get, can we look into improving the existing returned error, rather than setting up parallel "unrecognized version" errors in this package?

Reply:
can we fix like that:
```
 NewUnrecognizedVersionError(err.Error(), *gvk, data)
```